mne/tests/test_source_tfr.py

```diff
@@ -0,0 +1,201 @@
+from copy import deepcopy
+import os.path as op
+
+import numpy as np
+from numpy.testing import (assert_array_almost_equal, assert_array_equal,
+                           assert_allclose, assert_equal)
+import pytest
+
+from mne.utils import (_TempDir, requires_h5py, run_tests_if_main)
+from mne.source_tfr import SourceTFR
+
+
+def _fake_stfr():
+    """Create a fake SourceTFR."""
+    verts = [np.arange(10), np.arange(90)]
+    return SourceTFR(np.random.rand(100, 20, 10), verts, 0, 1e-1, 'foo')
+
+
+def _fake_kernel_stfr():
+    """Create a fake kernelized SourceTFR."""
+    kernel = np.random.rand(100, 40)
+    sens_data = np.random.rand(40, 20, 10)
+    verts = [np.arange(10), np.arange(90)]
+    return SourceTFR((kernel, sens_data), verts, 0, 1e-1, 'foo')
+
+
+def test_stfr_kernel_equality():
+    """Test kernelized and normal SourceTFR equality."""
+    kernel = np.random.rand(100, 40)
+    sens_data = np.random.rand(40, 10, 30)
+    verts = [np.arange(10), np.arange(90)]
+    data = np.tensordot(kernel, sens_data, axes=([-1], [0]))
+    tmin = 0
+    tstep = 1e-3
+
+    kernel_stfr = SourceTFR((kernel, sens_data), verts, tmin, tstep)
+    full_stfr = SourceTFR(data, verts, tmin, tstep)
+
+    # check if data is in correct shape
+    expected = [100, 10, 30]
+    assert_allclose([kernel_stfr.shape, full_stfr.shape,
+                     kernel_stfr.data.shape, full_stfr.data.shape],
+                    [expected] * 4)
+    assert_allclose(kernel_stfr.data, full_stfr.data)
+
+    # alternatively with the fake data
+    assert_equal(_fake_stfr().shape, _fake_kernel_stfr().shape)
+    assert_array_equal(_fake_stfr().data.shape, _fake_kernel_stfr().data.shape)
+
+
+def test_stfr_attributes():
+    """Test SourceTFR attributes."""
+    stfr = _fake_stfr()
+    stfr_kernel = _fake_kernel_stfr()
+
+    n_times = len(stfr.times)
+    assert_equal(stfr._data.shape[-1], n_times)
+    assert_array_equal(stfr.times, stfr.tmin + np.arange(n_times) * stfr.tstep)
+
+    assert_array_almost_equal(
+        stfr.times, [0., 0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9])
+
+    def attempt_times_mutation(stfr):
+        stfr.times -= 1
+
+    def attempt_assignment(stfr, attr, val):
+        setattr(stfr, attr, val)
+
+    # .times is read-only
+    pytest.raises(ValueError, attempt_times_mutation, stfr)
+    pytest.raises(ValueError, attempt_assignment, stfr, 'times', [1])
+
+    # Changing .tmin or .tstep re-computes .times
+    stfr.tmin = 1
+    assert (type(stfr.tmin) == float)
+    assert_array_almost_equal(
+        stfr.times, [1., 1.1, 1.2, 1.3, 1.4, 1.5, 1.6, 1.7, 1.8, 1.9])
+
+    stfr.tstep = 1
+    assert (type(stfr.tstep) == float)
+    assert_array_almost_equal(
+        stfr.times, [1., 2., 3., 4., 5., 6., 7., 8., 9., 10.])
+
+    # tstep <= 0 is not allowed
+    pytest.raises(ValueError, attempt_assignment, stfr, 'tstep', 0)
+    pytest.raises(ValueError, attempt_assignment, stfr, 'tstep', -1)
```

Comment:
I know this is compact, but when I read this I need to go from 
`attempt_assignment, stfr, 'tstep', 0` to `attempt_assignment(foo=foo, bar=bar..)` inside my head. Its better to write `with pytest.raises(ValueError, match='AND A MEANINGFUL ERROR MATCH THAT TELLS ME WHY THIS SHOULD FAIL')`